File: examples/cpp/composite.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class FileSystemComponent {
public:
    virtual ~FileSystemComponent() = default;
    virtual void display(const std::string& indent = "") const = 0;
    virtual int getSize() const = 0;
};

class File : public FileSystemComponent {
    std::string name_;
    int size_;
public:
    File(const std::string& name, int size) : name_(name), size_(size) {}
    void display(const std::string& indent = "") const override {
        std::cout << indent << "File: " << name_ << " (" << size_ << " KB)" << std::endl;
    }
    int getSize() const override { return size_; }
};

class Directory : public FileSystemComponent {
    std::string name_;
    std::vector<std::shared_ptr<FileSystemComponent>> children_;
public:
    Directory(const std::string& name) : name_(name) {}
    void add(std::shared_ptr<FileSystemComponent> comp) { children_.push_back(comp); }
    void display(const std::string& indent = "") const override {
        std::cout << indent << "Directory: " << name_ << " (" << getSize() << " KB)" << std::endl;
        for (const auto& child : children_) child->display(indent + "  ");
    }
    int getSize() const override {
        int total = 0;
        for (const auto& child : children_) total += child->getSize();
        return total;
    }
};

int main() {
    std::cout << "=== Composite: File System ===" << std::endl;

    auto root = std::make_shared<Directory>("root");
    auto docs = std::make_shared<Directory>("docs");
    auto src = std::make_shared<Directory>("src");

    docs->add(std::make_shared<File>("readme.md", 5));
    docs->add(std::make_shared<File>("guide.pdf", 120));
    src->add(std::make_shared<File>("main.cpp", 15));
    src->add(std::make_shared<File>("utils.cpp", 10));
    root->add(docs);
    root->add(src);
    root->add(std::make_shared<File>("Makefile", 2));

    root->display();

    return 0;
}

File: examples/csharp/Patterns/Composite.cs
namespace DesignPatterns.Composite
{
    public abstract class FileSystemEntry
    {
        public string Name { get; }
        protected FileSystemEntry(string name) => Name = name;
        public abstract int GetSize();
        public abstract void Display(string indent = "");
    }

    public class File : FileSystemEntry
    {
        private readonly int _size;
        public File(string name, int size) : base(name) => _size = size;
        public override int GetSize() => _size;
        public override void Display(string indent = "") =>
            Console.WriteLine($"{indent}{Name} ({_size} KB)");
    }

    public class Directory : FileSystemEntry
    {
        private readonly List<FileSystemEntry> _children = new();
        public Directory(string name) : base(name) { }

        public void Add(FileSystemEntry entry) => _children.Add(entry);

        public override int GetSize() => _children.Sum(c => c.GetSize());

        public override void Display(string indent = "")
        {
            Console.WriteLine($"{indent}{Name}/ ({GetSize()} KB)");
            foreach (var child in _children)
                child.Display(indent + "  ");
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var root = new Directory("root");
            root.Add(new File("readme.txt", 5));

            var src = new Directory("src");
            src.Add(new File("main.cs", 12));
            src.Add(new File("utils.cs", 8));
            root.Add(src);

            var docs = new Directory("docs");
            docs.Add(new File("guide.pdf", 120));
            root.Add(docs);

            root.Display();
        }
    }
}
